src/TurbulenceClosures/turbulence_closure_implementations/scale_invariant_smagorinsky.jl

```diff
@@ -77,30 +72,21 @@ end
     @inbounds νₑ[i, j, k] = cₛ² * (Δᶠ(i, j, k, grid))^2 * sqrt(2Σ²)
 end
 
-@inline directional_average!(averaged_field, field,   ::XDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=1)))
-@inline directional_average!(averaged_field, field,   ::YDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=2)))
-@inline directional_average!(averaged_field, field,   ::ZDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=3)))
-@inline directional_average!(averaged_field, field,  ::XYDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=(1, 2))))
-@inline directional_average!(averaged_field, field,  ::XZDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=(1, 3))))
-@inline directional_average!(averaged_field, field,  ::YZDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=(2, 3))))
-@inline directional_average!(averaged_field, field, ::XYZDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=(1, 2, 3))))
-
 function compute_diffusivities!(diffusivity_fields, closure::ScaleInvariantSmagorinsky, model; parameters = :xyz)
     arch = model.architecture
     grid = model.grid
     buoyancy = model.buoyancy
     velocities = model.velocities
     tracers = model.tracers
 
-    launch!(arch, grid, parameters, compute_LM_MM!,
-            diffusivity_fields.LM, diffusivity_fields.MM, grid, closure, velocities)
+    LM_op = KernelFunctionOperation{Center, Center, Center}(LᵢⱼMᵢⱼ_ccc, grid, velocities...)
+    MM_op = KernelFunctionOperation{Center, Center, Center}(MᵢⱼMᵢⱼ_ccc, grid, velocities...)
 
-    # TODO: optimize this
-    directional_average!(diffusivity_fields.LM_avg, diffusivity_fields.LM, closure.averaging)
-    directional_average!(diffusivity_fields.MM_avg, diffusivity_fields.MM, closure.averaging)
+    LM_avg = Field(Average(LM_op, dims=closure.averaging.dims))
+    MM_avg = Field(Average(MM_op, dims=closure.averaging.dims))
```

Comment:
you want to preallocate these in `DiffusivityFields`, and just `compute!` them in `compute_diffusivities!`.

Reply:
Yes, thanks. I realized yesterday that preallocating them in `DiffusivityFields` would require creating some new methods for `DiffusivityFields` in which we can pass `model`. So I was trying to see if it made sense to not modify the methods and instead do this. Surprisingly, I didn't see much of a performance decrease in this current version, although I've only tried for small models on CPUs.

I'll preallocate them in `DiffusivityFields` as you suggested though!

Reply:
> would require creating some new methods for DiffusivityFields in which we can pass model.

Why is that?

Reply:
Ah because they depend on velocities. That is a little puzzle...

One option is to pass `velocities` into `DiffusivityFields`. This is a fairly large change that might affect a lot of code, but could be a positive change perhaps. Maybe, we want to save that for after this PR.

You can also preallocate the field you're averaging into, but not the kernel function operation. Then you can compute the average with something like

```julia
mean!(field, kernel_function_operation)
```

(instead of `compute!`)

That could be a stop gap that will still let you do high performance simulation while we figure out the right pattern for this

Reply:
I'm not sure I follow your last suggestion. What I understood you're suggesting is something like the following, but which doesn't seem to work:

```julia
julia> op = Average(model.velocities.u^2, dims=(1, 2)) # Stand-in for the Lij and Mij functions
Average of BinaryOperation at (Face, Center, Center) over dims (1, 2)
└── operand: BinaryOperation at (Face, Center, Center)
    └── grid: 32×32×32 RectilinearGrid{Float64, Periodic, Periodic, Periodic} on CPU with 3×3×3 halo

julia> f = Field{Nothing, Nothing, Center}(grid)
1×1×32 Field{Nothing, Nothing, Center} reduced over dims = (1, 2) on RectilinearGrid on CPU
├── grid: 32×32×32 RectilinearGrid{Float64, Periodic, Periodic, Periodic} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Nothing, east: Nothing, south: Nothing, north: Nothing, bottom: Periodic, top: Periodic, immersed: ZeroFlux
└── data: 1×1×38 OffsetArray(::Array{Float64, 3}, 1:1, 1:1, -2:35) with eltype Float64 with indices 1:1×1:1×-2:35
    └── max=0.0, min=0.0, mean=0.0

julia> mean!(f, op)
ERROR: MethodError: no method matching mean!(::Field{…}, ::Reduction{…})

Closest candidates are:
  mean!(::Union{Oceananigans.Fields.AbstractField{Nothing}, Oceananigans.Fields.AbstractField{<:Any, Nothing}, Oceananigans.Fields.AbstractField{<:Any, <:Any, Nothing}}, ::AbstractArray; kwargs...)
   @ Oceananigans ~/repos/Oceananigans.jl2/src/Fields/field.jl:744
  mean!(::AbstractArray, ::AbstractArray)
   @ Statistics ~/bin/julia-1.10.2/share/julia/stdlib/v1.10/Statistics/src/Statistics.jl:140
```

Or maybe I did understand correctly, but we'd need to implement the relevant method for reductions?

Reply:
I guess you need `mean!(f, op.operand)`. Like here:

https://github.com/CliMA/Oceananigans.jl/blob/e7bbb1b77066f49a06c9f20942c70c2950de158b/src/Fields/scans.jl#L69

Reply:
Thanks! I think it works now